flake8_quotes/docstring_detection.py

```diff
@@ -0,0 +1,74 @@
+import tokenize
+
+# I don't think this is a minimized state machine, but it's clearer this
+# way. Namely, the class vs. function states can be merged
+
+# In the start of the module when we're expecting possibly a string that gets marked as a docstring
+STATE_EXPECT_MODULE_DOCSTRING = 0
+# After seeing the class keyword, we're waiting for the block colon (and do bracket counting)
+STATE_EXPECT_CLASS_COLON = 1
+# After seeing the colon in a class definition we're expecting possibly a docstring
+STATE_EXPECT_CLASS_DOCSTRING = 2
+# Same as EXPECT_CLASS_COLON, but for function definitions
+STATE_EXPECT_FUNCTION_COLON = 3
+# Same as EXPECT_CLASS_DOCSTRING, but for function definitions
+STATE_EXPECT_FUNCTION_DOCSTRING = 4
+# Just skipping tokens until we observe a class or a def.
+STATE_OTHER = 5
+
+# These tokens don't matter here - they don't get in the way of docstrings
+TOKENS_TO_IGNORE = [
+    tokenize.NEWLINE,
+    tokenize.INDENT,
+    tokenize.DEDENT,
+    tokenize.NL,
+    tokenize.COMMENT,
+]
+
+
+def get_docstring_tokens(tokens):
+    state = STATE_EXPECT_MODULE_DOCSTRING
+    # The number of currently open parentheses, square brackets, etc.
+    # This doesn't check if they're properly balanced, i.e. there isn't ([)], but we shouldn't
+    # need to - if they aren't, it shouldn't parse at all, so we ignore the bracket type
+    bracket_count = 0
+    docstring_tokens = set()
+
+    for token in tokens:
+        if token.type in TOKENS_TO_IGNORE:
+            continue
```

Comment:
Ah, I see. It might help to have a comment about that in the code. For example:

```python
# If we have a token that doesn't affect our state, then skip it
# DEV: We eagerly skip to prevent hitting accidental resets
if token.type in TOKENS_TO_IGNORE:
```